### PR TITLE
Fix #3305 and #3309

### LINKED ISSUE #3305: New REPL does not catch Exceptions
```scala
> console
scala> null.toString 
[error] (run-main-2) java.lang.reflect.InvocationTargetException
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at dotty.tools.repl.Rendering.valueOf(Rendering.scala:56)
	at dotty.tools.repl.Rendering.renderVal(Rendering.scala:75)
	at dotty.tools.repl.ReplDriver.displayMembers$5$$anonfun$4(ReplDriver.scala:265)
       ...
Caused by: java.lang.ExceptionInInitializerError
	at rs$line$1.res0Show(rs$line$1)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at dotty.tools.repl.Rendering.valueOf(Rendering.scala:56)
	...
Caused by: java.lang.NullPointerException
	at rs$line$1$.<init>(rs$line$1:1)
	at rs$line$1$.<clinit>(rs$line$1)
	at rs$line$1.res0Show(rs$line$1)
	...
[error] (compile:console) Nonzero exit code: 1
[error] Total time: 5 s, completed Oct 11, 2017 1:46:22 PM
> 
```

### PR DESCRIPTION
**Pull request by allanrenucci**

### CODE REVIEW

File: compiler/src/dotty/tools/repl/Rendering.scala [hunk 1]
```diff
@@ -70,10 +72,26 @@ private[repl] class Rendering(compiler: ReplCompiler,
...
+
+      resultValue.map(value => s"$dcl = $value")
+    }
+    catch { case ex: InvocationTargetException => Some(renderError(ex)) }
+  }
+
+  /** Render the stack trace of the underlying exception */
+  private def renderError(ex: InvocationTargetException): String = {
+    val cause = ex.getCause match {
+      case ex: ExceptionInInitializerError => ex.getCause
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/repl/Rendering.scala]: I am trying to isolate the underlying exception. This is based on nothing but experiment:
```scala
> null.toString
Exception in thread "main" java.lang.reflect.InvocationTargetException
	...
Caused by: java.lang.ExceptionInInitializerError
	...
Caused by: java.lang.NullPointerException
	...
```
```scala
scala> def foo: Int = foo + 1 
def foo: Int
scala> foo 
Exception in thread "main" java.lang.reflect.InvocationTargetException
	...
Caused by: java.lang.StackOverflowError
	...
```
I am sure there is a better a way to do it.

Comment (smarter) [reply compiler/src/dotty/tools/repl/Rendering.scala]: We really need tests for all of those :)

**Comment by allanrenucci** [comment]
In case of an exception, the result still get assigned to a value. Maybe this can be fixed in another PR.
```scala
scala> null.toString 
java.lang.NullPointerException
	at rs$line$1$.<init>(rs$line$1:1)
	at rs$line$1$.<clinit>(rs$line$1)
       ...
scala> res0 
java.lang.NoClassDefFoundError: Could not initialize class rs$line$1$
	at rs$line$2$.<init>(rs$line$2:1)
	at rs$line$2$.<clinit>(rs$line$2)
	at rs$line$2.res1Show(rs$line$2)
```

**Comment by smarter** [comment]
Also, the commit message titles could be improved to say what is being fixed, not just give the issue number.

File: compiler/src/dotty/tools/repl/Rendering.scala (hunk 1 above)
Comment (allanrenucci) [reply compiler/src/dotty/tools/repl/Rendering.scala]: Working on it

File: compiler/src/dotty/tools/repl/ReplCompiler.scala [hunk 2]
```diff
@@ -235,10 +235,14 @@ class ReplCompiler(val directory: AbstractFile) extends Compiler {
                             EmptyValDef,
                             state.imports.map(_._1) :+ valdef)
 
+        val pos =
+          if (trees.isEmpty) Position(0, expr.length)
+          else Position(trees.head.pos.start, trees.last.pos.end)
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/repl/ReplCompiler.scala]: @felixmulder Should the position always be `Position(0, expr.length)`?

Comment (felixmulder) [reply compiler/src/dotty/tools/repl/ReplCompiler.scala]: You're creating the wrapper here - so it needs to have a position that spans all the trees, looks equivalent :)

Comment (allanrenucci) [reply compiler/src/dotty/tools/repl/ReplCompiler.scala]: Then I would rather not special case

File: compiler/test-resources/repl/parsing [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+val res1: Int = 2
+scala> 1; 
+val res2: Int = 1
+scala> 1;; 2 
+val res3: Int = 1
+val res4: Int = 2
+scala> } 
+1 | }
+  | ^
+  | eof expected, but '}' found
```
Comment (allanrenucci) [inline compiler/test-resources/repl/parsing]: @felixmulder Why do I need a space after `|`, even though when I run the repl on my machine it doesn't print the space. Is it a bug in the testing framework?

Comment (felixmulder) [reply compiler/test-resources/repl/parsing]: Sounds like it!

**Pull request merged**